gen3_tracker/common/__init__.py

```diff
@@ -319,9 +320,42 @@ def create_id(resource, project_id) -> str:
     from gen3_tracker import ACED_NAMESPACE
     assert resource, "resource required"
     assert project_id, "project_id required"
-    # identifier string is not unique when it complies with FHIR server. Using something else
-    # identifier_string = identifier_to_string(resource.identifier)
-    return str(uuid.uuid5(ACED_NAMESPACE, f"{project_id}/{resource.resource_type}/{resource}"))
+    identifier_string = identifier_to_string(resource.identifier)
+    return str(uuid.uuid5(ACED_NAMESPACE, f"{project_id}/{resource.resource_type}/{identifier_string}"))
+
+
+def create_object_id(path: str, project_id: str) -> str:
+    """Create a unique did for this object within a project."""
+    from gen3_tracker import ACED_NAMESPACE
+
+    def _normalize_file_url(_path: str) -> str:
+        """Strip leading ./ and file:/// from file urls."""
+        _path = re.sub(r'^file:\/\/\/', '', _path)
+        _path = re.sub(r'^\.\/', '', _path)
+        return _path
+
+    base = f"{project_id}::{_normalize_file_url(path)}".encode()
+    object_id = str(uuid.uuid5(ACED_NAMESPACE, base))
+
+    return object_id
+
+
+def assert_valid_id(resource, project_id):
+    """Ensure that the id is correct."""
+    assert resource, "resource required"
+    assert project_id, "project_id required"
+    if resource.resource_type == "DocumentReference":
+        document_reference: DocumentReference = resource
+        official_identifier = document_reference.content[0].attachment.url
```

Comment:
Since url is not required in [Attachment](https://build.fhir.org/datatypes.html#Attachment), what is the official identifier if the field doesn't exist?

Separately, should we use `official_identifier` in line 350 since it's defined?

Reply:
For a document reference, we depend on it being there.
`The actual content of an Attachment can be conveyed directly using the data element or a URL reference can be provided. `

Reply:
> Separately, should we use official_identifier in line 350 since it's defined?

?

Reply:
got it so you do want it to throw an error if it's not there since we don't support pulling from the data element?

As for the using `official_identifier` point, you can do

```
# currently
recreate_id = create_object_id(document_reference.content[0].attachment.url, project_id)

# can change to 
recreate_id = create_object_id(official_identifier, project_id)
```